syntax = "proto3";

package qms;

option java_package = "com.nevmem.qms";
option java_outer_classname = "TicketProto";

import "proto/user.proto";


message Ticket {
    enum State {
        WAITING = 0;
        PROCESSING = 1;
        PROCESSED = 2;
    }

    enum Resolution {
        NONE = 0;  // ticket state != PROCESSED
        SERVICED = 1;  // User was serviced correctly
        NOT_SERVICED = 2;  // User wasn't serviced (not enough documents and etc.)
        KICKED = 3;  // User was kicked from queue by system before servicing
        GONE = 4;  // User left queue by himself before serving
    }

    string id = 1;

    string user_id = 2;
    string service_id = 3;
    string ticket_id = 4;
    string organization_id = 8;
    string window = 11;

    float enqueue_at = 5;
    float accepted_at = 9;

    User user = 7;
    State state = 10;
    Resolution resolution = 12;  // set after ticket processed
}

message TicketInfo {
    Ticket ticket = 1;

    uint32 remaining_time = 2;
    uint32 people_in_front_count = 3;
}


message TicketList {
    repeated Ticket tickets = 1;
}
